Implement cell cycle length tracking

pohaoc2 opened the pull request:
Resolves #27 
_Estimated size: xsmall_

Track cell cycle lengths by recording each cycle length every time a cell divides.

Summary of major changes:
- PatchCell currently includes `cycles` attribute of type `private final Bag` to record the lengths for each division
- Introduced `addCycle(double cycleLength)` and `getCycles()` methods in `PatchCell` to allow adding and retrieving cycle lengths.
- Initialized `start` value when a cell enters proliferative state. And each time the proliferative module is stepped, the start value increases by 1. 
- Removed `getStart()` and `getStop()` in the PatchModule because we no longer need these functions for modules in v3 to setup schedule()
- Added tests to verify both the size and values stored in `cycles`
- Need to output the cell cycles durations after we plan on how to save information

pohaoc2 commented:
Hi @jessicasyu, I've changed the way I record cycle lengths, and isolated test functions to test one function at a time. Please review the PR again, thank you!

cainja commented on src/arcade/patch/agent/module/PatchModuleProliferation.java:
```diff
@@ -50,7 +50,7 @@ public PatchModuleProliferation(PatchCell cell) {
         targetVolume = 2 * cell.getCriticalVolume();
         maxHeight = cell.getCriticalHeight();
-
+        start = 0; // start ticking at the time of the creation of the module
```
can we call this duration rather than start?

cainja replied:
I also don't think we need the comment

jessicasyu commented on src/arcade/patch/agent/module/PatchModule.java:
```diff
@@ -17,7 +17,7 @@ public abstract class PatchModule implements Module {
+    int start;
 
     /** Tick the {@code Module} was stopped. */
     double stop;
```
You can delete `stop` since it is no longer used

jessicasyu commented on src/arcade/patch/agent/module/PatchModule.java:
```diff
@@ -17,7 +17,7 @@ public abstract class PatchModule implements Module {
 
     /** Tick the {@code Module} was started. */
-    double start;
+    int start;
```
Move this variable into `PatchModuleProliferation`, since it is only used by `PatchModuleProliferation`

jessicasyu commented on test/arcade/patch/agent/cell/PatchCellRandomTest.java:
```diff
@@ -64,6 +64,37 @@ public static void setupMocks() {
         doReturn(0).when(parametersMock).getInt(any(String.class));
     }
 
+    @Test
```
I think these tests should be in `PatchCellTest` and not `PatchCellRandomTest`

pohaoc2 replied:
create `PatchCellTest.java` and move tests to this file

jessicasyu commented on test/arcade/patch/agent/cell/PatchCellRandomTest.java:
```diff
@@ -64,6 +64,37 @@ public static void setupMocks() {
+    @Test
+    public void getCycles_beforeDivision_returnsEmpty() {
+        PatchCellRandom cell = new PatchCellRandom(baseContainer, locationMock, parametersMock);
+        // Test the values in the cell.cycles bag
```
You don't need these comments -- the purpose of the test should be clear from the test name

pohaoc2 replied:
Deleted.

jessicasyu commented on test/arcade/patch/agent/cell/PatchCellRandomTest.java:
```diff
@@ -64,6 +64,37 @@ public static void setupMocks() {
     }
 
+    @Test
+    public void getCycles_beforeDivision_returnsEmpty() {
```
What about "after division"? If you have a test for a "before" condition, you generally want to test the "after" condition as well

pohaoc2 replied:
I agree it doesn't make sense to have only the "before" scenario and no "after". I think the "after" condition is tested in the addCycle_givenCycle_appendBagSize(), so I changed it to `whenInitizeCell`

jessicasyu commented on test/arcade/patch/agent/cell/PatchCellRandomTest.java:
```diff
@@ -64,6 +64,37 @@ public static void setupMocks() {
+    }
+
+    @Test
+    public void addCycle_givenCycle_appendBagSize() {
```
I think this test is redundant with `addCycle_givenCycles_appendValues`. You can add the assert for size in the other test. Generally, tests for equality don't need separate tests for size and content

pohaoc2 replied:
Combine tests for size and values

pohaoc2 commented:
@cainja, thank you for the comments, I have modified the variable name and remove comment.
@jessicasyu, I have 1) combined the duplicate tests, 2) created `PatchCellTest` and move these tests to this file, and 3) changed the function name `beforeDivision` to `whenCellInitialized` because I test `afterDivision` scenario in `addCycle_givenCycles` test. Thank you for the review!

cainja commented on test/arcade/patch/agent/cell/PatchCellTest.java:
```diff
@@ -0,0 +1,80 @@
+
+    @Test
+    public void getCycles_whenCellInitialized_returnsEmpty() {
+        PatchCellRandom cell = new PatchCellRandom(baseContainer, locationMock, parametersMock);
```
`PatchCellTest` shouldn't be testing `PatchCellRandom`

pohaoc2 replied:
Thank you for pointing out that `PatchCellRandom` is created in the `PatchCellTest`. In the `PatchCellTest`, I use `PatchCellMock` to create instances of `PatchCell`. Since I am new to writing tests, I followed a similar approach as how `PottsCellMock` is created in `PottsCellTest`. Could you please let me know if this method is appropriate or if there’s a better way to handle this? Thank you!

pohaoc2 commented on test/arcade/patch/agent/cell/PatchCellTest.java:
```diff
@@ -0,0 +1,105 @@
+    public static void setupMocks() {
+        locationMock = mock(PatchLocation.class);
+        parametersMock = spy(new Parameters(new MiniBox(), null, null));
+        doReturn(0.0).when(parametersMock).getDouble(any(String.class));
```
Hi @cainja, I moved the mock behaviors into each test. Please review again, thank you!

jessicasyu commented:
@pohaoc2 Please change the base branch to `main` and make sure you have those changes merged in

pohaoc2 commented:
Hi, @jessicasyu, I merged `main` branch to my branch, and change the base branch `main`.
